### DISCUSSION TITLE
More official avrdude 7.3 binaries?

### DISCUSSION THREAD

**Discussion by mcuee** [avrdudes/avrdude#1675] [General]
Now we only provide Windows MSVC build binaries with avrdude 7.3 release here.
https://github.com/avrdudes/avrdude/releases/tag/v7.3

This is in line with avrdude 7.0/7.1/7.2 release.

However, Windows MSVC build has some limitations and sometimes the user may need to use the mingw32/mingw64 build.
https://github.com/avrdudes/avrdude/wiki/Getting-Nightly-Builds-for-AVRDUDE

And Arduino Core distributions traditionally like the static linking binaries for Linux, macOS and Windows mingw32 build using the Arduino arvdude-packing cross-build.

**Comment by mcuee** [comment]
@stefanrueger @MCUdude and @dl8dtl 

Do you think it is a good idea to publish mingw32/mingw64 build as official Windows binaries as well?

Currently the deploy github action only upload Windows MSVC binaries of an official release.
https://github.com/avrdudes/avrdude/blob/main/.github/workflows/deploy.yml
https://github.com/avrdudes/avrdude/actions/runs/7813933855

The above deploy github action does build MSYS2 mingw32/mingw64/clang32/clang64/ucrt64 binaries for avrdude 7.3 release. So we can decide whether to upload them or not.

**Reply by mcuee** [reply]
@stefanrueger @MCUdude and @dl8dtl

My suggestion is to upload at least MSYS2 mingw64 and mingw32 binaries. Please comment. Thanks.

https://github.com/avrdudes/avrdude/wiki/Getting-Nightly-Builds-for-AVRDUDE
**Known limitation for the above MSVC build**

1. https://github.com/avrdudes/avrdude/issues/1552 (no libserialport support)
2. https://github.com/avrdudes/avrdude/issues/1253 (slow speed when using arduino-ft232r or similar programmer)
3. https://github.com/avrdudes/avrdude/issues/968 (not working with USB Composite Device programer like USBASP with the FW from https://github.com/dioannidis/usbasp)
4. https://github.com/avrdudes/avrdude/discussions/1440 (no support of CH341A)

**Reply by mcuee** [reply]
@stefanrueger @MCUdude and @dl8dtl

Any objections?

**Reply by mcuee** [reply]
@stefanrueger @MCUdude and @dl8dtl
I have taken the liberty and added the mingw64 and mingw32 binaries to the avrdude 7.3 release page.
https://github.com/avrdudes/avrdude/releases/tag/v7.3

**Reply by mcuee** [reply]
@MCUdude and @stefanrueger and @dl8dtl 

I have added the notes for Windows binaries here. Please check if the wordings need changes or not. Thanks.
https://github.com/avrdudes/avrdude/releases/tag/v7.3

**Reply by mcuee** [reply]
Linux static linked binaries (with libserialport support) uploaded to the 7.3 release page.
https://github.com/avrdudes/avrdude/releases/tag/v7.3

**Comment by mcuee** [comment]
As for the binaries generated by arduino-packing script, as of now, the script does not deal with releases yet.So the version number will not matching -- one commit after the release tag. Probably not a real problem though.
https://github.com/avrdudes/avrdude/blob/main/.github/workflows/arduino_packing.yml
https://github.com/avrdudes/avrdude/actions/runs/7814004029
```
mcuee@UbuntuSwift3 ~/build/avr/avrdude_bin/avrdude_Linux_32bit/bin $ ./avrdude 
Usage: avrdude [options]
Options:
  -p <partno>            Specify AVR device; -p ? lists all known parts
  -p <wildcard>/<flags>  Run developer options for matched AVR devices,
                         e.g., -p ATmega328P/s or /S for part definition
  -b <baudrate>          Override RS-232 baud rate
  -B <bitclock>          Specify bit clock period (us)
  -C <config-file>       Specify location of configuration file
  -C +<config-file>      Specify additional config file, can be repeated
  -N                     Do not load config file 
  -c <programmer>        Specify programmer; -c ? and -c ?type list all
  -c <wildcard>/<flags>  Run developer options for matched programmers,
                         e.g., -c 'ur*'/s for programmer info/definition
  -A                     Disable trailing-0xff removal for file/AVR read
  -D                     Disable auto erase for flash memory; implies -A
  -i <delay>             ISP Clock Delay [in microseconds]
  -P <port>              Connection; -P ?s or -P ?sa lists serial ones
  -r                     Reconnect to -P port after "touching" it; wait
                         400 ms for each -r; needed for some USB boards
  -F                     Override invalid signature or initial checks
  -e                     Perform a chip erase
  -O                     Perform RC oscillator calibration (see AVR053)
  -t                     Run an interactive terminal when it is its turn
  -T <terminal cmd line> Run terminal line when it is its turn
  -U <memstr>:r|w|v:<filename>[:format]
                         Carry out memory operation when it is its turn
                         Multiple -t, -T and -U options can be specified
  -n                     Do not write to the device whilst processing -U
  -V                     Do not automatically verify during -U
  -E <exitsp>[,<exitsp>] List programmer exit specifications
  -x <extended_param>    Pass <extended_param> to programmer, see -xhelp
  -v                     Verbose output; -v -v for more
  -q                     Quell progress output; -q -q for less
  -l logfile             Use logfile rather than stderr for diagnostics
  -?                     Display this usage

avrdude version 7.3-20240208 (0179026), https://github.com/avrdudes/avrdude
```

**Reply by mcuee** [reply]
@MCUdude 

Do you consider the version number a real problem for your Arduino Core distributions?

**Reply by MCUdude** [reply]
I'd argue that the ideal name for a released version of Avrdude would _just_ be 7.3. Arduino adds their prefix to that, and that's fine.
But IMO, the date (and hash) should only be present on development versions where these things matters

**Reply by mcuee** [reply]
@MCUdude 
So we need to improve the github action script to be able to build with tagged release.

**Reply by mcuee** [reply]
@umbynos 
The [avrdude-packing script in this repo](https://github.com/avrdudes/avrdude/blob/main/.github/workflows/arduino_packing.yml) is a cut-down version of[ your more detailed script.](https://github.com/arduino/avrdude-packing/blob/main/.github/workflows/release.yml) One missing feature is to deal with the tagged release (say 7.3 release). Just wondering if you can help here. Thanks.
Just wondering if you can help here. Thanks.

This was also mentioned previously.
* https://github.com/avrdudes/avrdude/discussions/1613#discussioncomment-8062857

**Reply by mcuee** [reply]
@MCUdude 
In the mean time, you can also just use the container and build the binaries by yourself. 

I will try to do that over the weekend if you think it is too troublesome for you.

**Reply by mcuee** [reply]
I have archived the arduino-packing github action binaries here.
https://github.com/mcuee/avrdude/releases/tag/v7.3

Technically the binaries are not the same as avrdude 7.3 release, but [one git commit ](https://github.com/avrdudes/avrdude/commit/0179026e09d19833e7dea06ed8db19f971138e4b ) after 7.3 release. Therefore the binaries will print  `avrdude version 7.3-20240208 (0179026)`.
https://github.com/avrdudes/avrdude/blob/v7.3/.github/workflows/arduino_packing.yml
https://github.com/avrdudes/avrdude/actions/runs/7814004029

**Comment by mcuee** [comment]
For arduino_packing binaries, here is the discussion.
* https://github.com/arduino/avrdude-packing/issues/31

Except macOS, I can build the binaries with libserialport support with the Arduino crossbuild container. The binaries (no macOS) are archived here.
https://github.com/mcuee/avrdude/releases/tag/v7.3

**Reply by mcuee** [reply]
@MCUdude @stefanrueger and @dl8dtl 

Just wondering if you think it is a good idea to upload the above-mentioned Linux binaries to avrude github release here as well? They do support libserialport.
https://github.com/avrdudes/avrdude/releases/tag/v7.3

```
mcuee@debian12vmn100:~/build/avrdude_linux_x64$ ./avrdude -c jtag2updi -P ch340 -p m4808 -t -v

avrdude: Version 7.3
         Copyright the AVRDUDE authors;
         see https://github.com/avrdudes/avrdude/blob/main/AUTHORS

         System wide configuration file is /home/mcuee/build/avrdude_linux_x64/avrdude.conf
         User configuration file is /home/mcuee/.avrduderc
         User configuration file does not exist or is not a regular file, skipping

         Using port            : /dev/ttyUSB0
         Using programmer      : jtag2updi
         Programmer baud rate  : 115200
         AVR Part              : ATmega4808
         Programming modes     : UPDI, SPM
         Programmer Type       : JTAGMKII_UPDI
         Description           : JTAGv2 to UPDI bridge
         Main MCU HW version   : 1
         Main MCU FW version   : 6.00
         Sec. MCU HW version   : 1
         Sec. MCU FW version   : 6.00
         Serial number         : 0a:0b:0c:0d:0e:0f
avrdude: silicon revision: 0.1
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e9650 (probably m4808)

avrdude: processing -t interactive terminal
avrdude> dump fuses
>>> dump fuses 0x0 0xa
Reading | ################################################## | 100% 0.04 s
0000  00 54 01 ff 00 c8 06 00  02 ff                    |.T........      |
avrdude> quit

avrdude done.  Thank you.
```

**Comment by mcuee** [comment]
I will close this discussion here.

More discussions will be in the following places.
* https://github.com/arduino/avrdude-packing/issues/31
* https://github.com/avrdudes/avrdude/issues/1679

**Reply by mcuee** [reply]
Re-open the discussuon as there are some confusions about which Windows binary to use.
* https://github.com/avrdudes/avrdude/discussions/1696

**Comment by MCUdude** [comment]
In my opinion, we could provide statically built binaries for those who want a binary that "just works", and dynamically linked ones for those who want to deal with dependencies. At the moment, I'm recommending the ones compiled by Arduino for those who just need a working binary for their system.

**Reply by mcuee** [reply]
It may be a bit difficult to provide dynamic linked binaries to be suitable for use for different Linux distros and macOS versions. We will rely on Linux distros and macOS Homebrew project to do that. People can also build from source. As of now, there is no official facility to build Windows dynamic linked binaries.

 I will upload the Linux static linked binaries (with libserialport support) to the 7.3 release page.
https://github.com/avrdudes/avrdude/releases/tag/v7.3

**Reply by mcuee** [reply]
Done, Linux binaries uploaded.
https://github.com/avrdudes/avrdude/releases/tag/v7.3
